rclpy/rclpy/parameter_client.py

```diff
@@ -0,0 +1,310 @@
+            self._get_parameter_client.service_is_ready(),
+            self._get_parameter_types_client.service_is_ready(),
+            self._describe_parameters_client.service_is_ready(),
+            self._set_parameters_atomically_client.service_is_ready(),
+        ])
+
+    def wait_for_service(self, timeout_sec: Optional[float] = None) -> bool:
+        """
+        Wait for all parameter services to be available.
+
+        :param timeout_sec: Seconds to wait. If ``None``, then wait forever.
+        :type timeout_sec: Union[float, None]
+        :return: ``True`` if all services were waite , ``False`` otherwise.
+        """
+        client_wait_fns = [
+                self._list_parameter_client.wait_for_service,
+                self._set_parameter_client.wait_for_service,
+                self._get_parameter_client.wait_for_service,
+                self._get_parameter_types_client.wait_for_service,
+                self._describe_parameters_client.wait_for_service,
+                self._set_parameters_atomically_client.wait_for_service,
+        ]
+
+        if timeout_sec is None:
+            return all([fn() for fn in client_wait_fns])
+
+        prev = time.time()
+        for wait_for_service in client_wait_fns:
+            if timeout_sec < 0 or not wait_for_service(timeout_sec):
+                return False
+            timeout_sec -= time.time() - prev
+            prev = time.time()
+        return True
+
+    def list_parameters(
+        self,
+        prefixes: Optional[List[str]] = None,
+        depth: int = 1,
+        callback: Optional[Callable] = None
+    ) -> Future:
+        """
+        List all parameters with given prefixs.
+
+        :param prefixes: List of prefixes to filter by.
+        :param depth: Depth of the parameter tree to list.
+        :param callback: Callback function to call when the request is complete.
+        :return: ``Future`` with the result of the request.
+        """
+        request = ListParameters.Request()
+        if prefixes:
+            request.prefixes = prefixes
+        request.depth = depth
+        future = self._list_parameter_client.call_async(request)
+        if callback:
+            future.add_done_callback(callback)
+        return future
+
+    def get_parameters(self, names: List[str], callback: Optional[Callable] = None) -> Future:
+        """
+        Get parameters given names.
+
+        :param names: List of parameter names to get.
+        :param callback: Callback function to call when the request is complete.
+        :return: ``Future`` with the result of the request.
+        """
+        request = GetParameters.Request()
+        request.names = names
+        future = self._get_parameter_client.call_async(request)
+        if callback:
+            future.add_done_callback(callback)
+        return future
+
+    def set_parameters(
+        self,
+        parameters: Sequence[Parameter],
```

Comment:
I think this type annotation is wrong. Looks like we're expecting a message type (`rcl_interfaces.msg.Parameter`), and not `rclpy.parameter.Parameter`. Although, maybe it's more convenient to accept an rclpy `Parameter` instead of (or in addition to) the messsage type. For example, our call in the test could change like,

```diff
-        future = self.client.set_parameters([
-            Parameter('int_param', Parameter.Type.INTEGER, 88).to_parameter_msg(),
-            Parameter('string/param', Parameter.Type.STRING, 'hello world').to_parameter_msg(),
-        ])
+        future = self.client.set_parameters([
+            Parameter('int_param', Parameter.Type.INTEGER, 88),
+            Parameter('string/param', Parameter.Type.STRING, 'hello world'),
+        ])
```

And we could do the conversion `to_parameter_msg()` inside this function as-needed.

Reply:
In this file `Parameter` refers to `rcl_interfaces.msg.Parameter`. But also makes sense to accept both and convert as-needed.